File: rust/src/detect.rs
use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct DetectedTool {
    pub tool: String,
    pub config_dir: PathBuf,
    pub display_name: String,
}

struct ToolSignature {
    tool: &'static str,
    dir_name: &'static str,
    display_name: &'static str,
    marker_files: &'static [&'static str],
}

const TOOL_SIGNATURES: &[ToolSignature] = &[
    ToolSignature {
        tool: "claude",
        dir_name: ".claude",
        display_name: "Claude Code",
        marker_files: &[".credentials.json", "settings.json"],
    },
    ToolSignature {
        tool: "gemini",
        dir_name: ".gemini",
        display_name: "Gemini CLI",
        marker_files: &["config.json", "settings.json"],
    },
    ToolSignature {
        tool: "codex",
        dir_name: ".codex",
        display_name: "Codex CLI",
        marker_files: &["config.json", "settings.json"],
    },
];

/// Detect installed agent tool configs on the system.
pub fn detect_tool_configs() -> Vec<DetectedTool> {
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
    let mut detected = Vec::new();

    for sig in TOOL_SIGNATURES {
        let config_dir = home.join(sig.dir_name);

        if !config_dir.exists() {
            continue;
        }

        let has_marker = sig
            .marker_files
            .iter()
            .any(|f| config_dir.join(f).exists());

        if has_marker {
            detected.push(DetectedTool {
                tool: sig.tool.to_string(),
                config_dir,
                display_name: sig.display_name.to_string(),
            });
        }
    }

    detected
}

File: src/detect.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

export interface DetectedTool {
  tool: string;
  configDir: string;
  displayName: string;
}

/**
 * Known tool config locations relative to the user's home directory.
 * Each entry lists candidate files that, if present, confirm the tool is installed.
 */
const TOOL_SIGNATURES: Array<{
  tool: string;
  dirName: string;
  displayName: string;
  markerFiles: string[];
}> = [
  {
    tool: "claude",
    dirName: ".claude",
    displayName: "Claude Code",
    markerFiles: [".credentials.json", "settings.json"],
  },
  {
    tool: "gemini",
    dirName: ".gemini",
    displayName: "Gemini CLI",
    markerFiles: ["config.json", "settings.json"],
  },
  {
    tool: "codex",
    dirName: ".codex",
    displayName: "Codex CLI",
    markerFiles: ["config.json", "config.toml", "auth.json", "settings.json"],
  },
];

/**
 * Detect installed agent tool configs on the system.
 *
 * For each known tool, checks whether its config directory exists and contains
 * at least one recognised marker file.  Returns an array of detected tools,
 * ordered Claude -> Gemini -> Codex.
 */
export function detectToolConfigs(): DetectedTool[] {
  const home = os.homedir();
  const detected: DetectedTool[] = [];

  for (const sig of TOOL_SIGNATURES) {
    const configDir = path.join(home, sig.dirName);

    if (!fs.existsSync(configDir)) continue;

    const hasMarker = sig.markerFiles.some((file) =>
      fs.existsSync(path.join(configDir, file))
    );

    if (hasMarker) {
      detected.push({
        tool: sig.tool,
        configDir,
        displayName: sig.displayName,
      });
    }
  }

  return detected;
}
